File: pset6/credit/credit.py
from cs50 import get_int
import math


def main():
    # prompt the user a credit card number CC
    CC = 0
    while (CC <= 0):
        CC = get_int("Number: ")

    # convert CC to string to obtain the length of number
    CC_string = str(CC)
    length_of_CC = len(CC_string)

    #validity = check_validity(length_of_CC, CC)
    # Considering CC has a valid number by Luhn’s Algorithm
    validity = True

    if validity == True:
        card_type = check_card_type(CC_string)
        print(card_type)
    else:
        print("INVALID")


# Extra Work: check if the number is valid by Luhn’s Algorithm
'''
def check_validity(length_of_CC, CC):
    sum1, sum2, total, digit_check = 0, 0, 0, False
    for i in range(length_of_CC + 1):

        if digit_check == False:
            sum1 = sum1 + (CC % 10)
            CC = math.floor(CC / 10 )
            digit_check = True

        elif digit_check == True:
            temp = 2 * (CC % 10)
            if temp >= 10:
                sum2 = sum2 + round(temp % 10 + temp / 10)
            else:
                sum2 = sum2 + temp

            CC = math.floor(CC / 10 )
            digit_check = False

    total = sum1 + sum2
    if total % 10 == 0:
        return True

    else:
        return False
'''


def check_card_type(CC_string):

    # obtain firts digit and second digit to check the type of card
    First_digit = int(CC_string[0])
    First_2_digits = int(CC_string[0]) * 10 + int(CC_string[1])

    # American Express
    print(First_digit, First_2_digits)
    if First_2_digits in [34, 37]:
        return "AMEX"

    # MASTERCARD
    elif First_2_digits in [51, 52, 53, 54, 55]:
        return "MASTERCARD"

    # VISA Card
    elif First_digit == 4:
        return "VISA"

    else:
        return "INVALID"


if __name__ == "__main__":
    main()

File: pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>
#include <math.h>

int main(void)
{

    // get the credit card number
    long card;
    do
    {
        card = get_long("Number: ");
    }
    while (card <= 0);



    // save the original number as a reference
    long original_card = card;

    // index: refers to the position of digit
    // sum1 = second last digit sequence
    // sum2 = last digit sequence

    int index = 0, sum1 = 0, sum2 = 0, add;
    while (card != 0)
    {
        // Addition of second last digit sequene
        if (index % 2 != 0)
        {
            add = card % 10;
            add = add * 2;
            if (add >= 10)
            {
                sum1 = sum1 + (add % 10 + add / 10);
            }
            else
            {
                sum1 = sum1 + add;
            }
            card = card / 10;
            index++;


        }

        // Addition of  last digit sequene
        else
        {
            add = card % 10;
            sum2 = sum2 + add;
            card = card / 10;
            index++;
        }

    }

    // total sum of both sequence
    int total = sum1 + sum2;

    // if the total Multiples of 10 ==> valid card
    if (total % 10 == 0)
    {

        // American Express validation check
        int start_amex = original_card / pow(10, index - 2);
        if ((index == 15) && (start_amex == 34 || start_amex == 37))
        {
            printf("AMEX\n");
            return 0;
        }


        // MasterCard validation check

        int start_master = original_card / pow(10, index - 2);
        if ((index == 16) && (start_master >= 51 && start_master <= 55))
        {
            printf("MASTERCARD\n");
            return 0;
        }

        // VISA validation check
        int start_visa = original_card / pow(10, index - 1) ;
        if ((index >= 13 && index <= 16) && (start_visa == 4))
        {
            printf("VISA\n");
            return 0;
        }


        printf("INVALID\n");

    }


    // if the total not one of Multiples of 10 ==> invalid card
    else
    {
        printf("INVALID\n");
    }

}

